design-proposals/klip-30-lambda-functions.md

```diff
@@ -0,0 +1,125 @@
+
+* `transform_keys(map, k => k)` - Applies the given lambda function to each key in the input `Map`, returning a new `Map` containing the transformed keys.
+
+* `transform_values(map, v => v)` - Applies the given lambda function to each value in the input `Map`, returning a new `Map` containing the transformed values.
+
+* `transform_map(map, (k, v) => new_k, (k, v) => new_v)` - Applies the given lambda functions to the keys and values of the input `Map`, respectively. A new `Map` containing the transformed keys and values is returned.
+
+* `filter_array(arr, x => bool)` - Filters the input `Array` using the given lambda function. A new `Array` is returned, containing only values for which the lambda function evaluated to `true`.
+
+* `filter_map(map, (k, v) => bool)` - Filters the input `Map` using the given lambda function. A new `Map` is returned, containing only the key-value pairs for which the lambda function evaluated to `true`.
+
+* `reduce_array(arr, s, (x, s) => s)` - Reduces the input `Array` down to a single value. `s` is the initial state and is passed into the scope of the lambda function. Each invocation returns a new value for `s`, which the next invocation will receive. `reduce_array` will return the final value of `s`.
+
+* `reduce_map(map, s, (k, v, s) => s)` - Reduces the input `Map` down to a single value. `s` is the initial state and is passed into the scope of the lambda function. Each invocation returns a new value for `s`, which the next invocation will receive. `reduce_map` will return the final value of `s`.
+
+## Documentation updates
+
+We must document and provide examples for the two core aspects of lambda functions:
+
+1. Lambda function syntax
+2. Invocation functions
+
+## Compatibility implications
+
+Lambda functions would be additive to ksqlDB's grammar and should therefore not introduce any incompatibilities into the grammar or public APIs. However, it is not currently known if a lambda function implementation would require backward-incompatible changes to ksqlDB's internals.
+
+## Security implications
+
+Since lambda functions would only allow SQL expressions that can already be used within queries, no new security implications should be introduced.
+
+## Open questions
+
+~~**1. Would it be possible to allow aggregates in lambdas within the context of a particular invocation function? For example,**~~
+
+~~- `agg_array(arr, x => sum(x))`~~
+
+~~If this is feasible then it would likely be worth doing. Without support for aggregates, users will not be able to easily perform any aggregation over a collection using `reduce_*`.~~
+
+**2. Would allowing lambda functions to be applied to types other than `Array` or `Map` provide any value?**
+
+**3. Will the lambda function implementation require backward-incompatible changes to ksqlDB's internals?**
+
+~~**4. Should we extend ksqlDB's UDF interface to allow users to write their own lambda invocation functions?**~~
```

Comment:
If we allow own lambda invocation functions, how that would impact our existing new query creation process? Would that require an extra step of loading user jars e.g.?